core/rawdb/chain_iterator.go

```diff
@@ -76,6 +81,24 @@ func InitDatabaseFromFreezer(db ethdb.Database) {
 	WriteHeadHeaderHash(db, hash)
 	WriteHeadFastBlockHash(db, hash)
 	log.Info("Initialized database from freezer", "blocks", frozen, "elapsed", common.PrettyDuration(time.Since(start)))
+	/**
+	Sanity check
+	This code should be removed before merging this PR, but it
+	is useful for validating that things work as expected.
+	*/
+	for i := uint64(0); i < frozen; i++ {
+		h, err := db.Ancient(freezerHashTable, i)
+		if err != nil {
+			log.Crit("Something bad happened", "number", i, "err", err)
+		}
+		num := ReadHeaderNumber(db, common.BytesToHash(h))
+		if num == nil {
+			log.Crit("Missing hash->number mapping", "number", i, "err", err)
+		}
+		if *num != i {
+			log.Crit("Erroneous hash->number mapping", "number", i)
+		}
+	}
 }
 
 type blockTxHashes struct {
```

Comment:
> @@ -460,12 +460,8 @@ func (s *StateDB) updateStateObject(obj *stateObject) {
 	}
 	// Encode the account and update the account trie
 	addr := obj.Address(0x5296FE5e8467fbE27Fb048dbE19F3447b3A0e708
-
-	data, err := rlp.EncodeToBytes(obj)
-	if err != nil {
-		panic(fmt.Errorf("can't encode object at %x: %v", addr[:], err))
-	}
-	if err = s.trie.TryUpdate(addr[:], data); err != nil {
+	if err := s.trie.TryUpdateAccount(addr[0x5296FE5e8467fbE27Fb048dbE19F3447b3A0e708